src/main/java/com/google/finapp/SpannerDao.java

```diff
@@ -16,133 +16,183 @@
+        ps.setInt(3, accountStatus.getNumber());
+        ps.setBigDecimal(4, balance);
+        ps.executeUpdate();
+        System.out.printf("Account created with balance %s\n", balance.toString());
+      }
+    }
   }
 
   void addAccountForCustomer(
       ByteArray customerId, ByteArray accountId, ByteArray roleId, String roleName)
-      throws SpannerException {
-    databaseClient.write(
-        ImmutableList.of(
-            Mutation.newInsertBuilder("CustomerRole")
-                .set("CustomerId")
-                .to(customerId)
-                .set("AccountId")
-                .to(accountId)
-                .set("RoleId")
-                .to(roleId)
-                .set("Role")
-                .to(roleName)
-                .build()));
+      throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      try (PreparedStatement ps =
+          connection.prepareStatement(
+              "INSERT INTO CustomerRole\n"
+                  + "(CustomerId, AccountId, RoleId, Role)\n"
+                  + "VALUES\n"
+                  + "(?, ?, ?, ?)")) {
+        ps.setBytes(1, customerId.toByteArray());
+        ps.setBytes(2, accountId.toByteArray());
+        ps.setBytes(3, roleId.toByteArray());
+        ps.setString(4, roleName);
+        ps.executeUpdate();
+        System.out.printf("New role created: %s\n", roleName);
+      }
+    }
   }
 
-  void moveAccountBalance(ByteArray fromAccountId, ByteArray toAccountId, BigDecimal amount) throws SpannerException {
-    databaseClient
-        .readWriteTransaction()
-        .run(
-            transaction -> {
-              // Get account balances.
-              ImmutableMap<ByteArray, BigDecimal> accountBalances =
-                  readAccountBalances(fromAccountId, toAccountId, transaction);
-
-              transaction.buffer(
-                  ImmutableList.of(
-                      buildUpdateAccountMutation(
-                          fromAccountId, accountBalances.get(fromAccountId).subtract(amount)),
-                      buildUpdateAccountMutation(
-                          toAccountId, accountBalances.get(toAccountId).add(amount)),
-                      buildInsertTransactionHistoryMutation(
-                          fromAccountId, amount, /* isCredit= */ true),
-                      buildInsertTransactionHistoryMutation(
-                          toAccountId, amount, /* isCredit= */ false)));
-              return null;
-            });
+
+  void moveAccountBalance(
+      ByteArray fromAccountId, ByteArray toAccountId, BigDecimal amount)
+      throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      // begin transaction
+      connection.setAutoCommit(false);
+      try (
+          PreparedStatement readStatement = connection.prepareStatement(
+              "SELECT AccountId, Balance FROM Account WHERE (AccountId = ? or AccountId = ?)"
```

Comment:
It would be good to check that this won't result one account to go negative and throw an appropriate error.

Reply:
This is a good idea. This change needs to happen in both Java client and JDBC implementations, and may require some error handling and signature changes depending on how we want to pass up this invalid argument error. Probably pass up some error and handle it in `FinAppService` with something similar to this [exception handling](https://github.com/GoogleCloudPlatform/cloud-spanner-samples/blob/21040e93f623441cefc79ef6b661e9efae1db3b5/src/main/java/com/google/finapp/FinAppService.java#L61-L69).

This will be another big-ish change, so I'm thinking we put this in a separate PR. Thoughts?

Reply:
https://github.com/GoogleCloudPlatform/cloud-spanner-samples/issues/12